src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -5902,6 +5902,9 @@ void ThreadSuspend::RestartEE(BOOL bFinishedGC, BOOL SuspendSucceded)
     // Flush the store buffers on all CPUs, to ensure that they all see changes made
     // by the GC threads. This only matters on weak memory ordered processors as 
     // the strong memory ordered processors wouldn't have reordered the relevant reads.
+    // This is needed to synchronize threads that were running in preemptive mode while
+    // the runtime was suspended and that will return to cooperative mode after the runtime 
```

Comment:
Is this because transition to coop does not do a read fence?

Reply:
I mean - do we need another Proc-wide fence between SuspendEE and starting GC? Transition to preemptive is semantically a release of the managed heap lock, but there are no fences.

Reply:
> Is this because transition to coop does not do a read fence?

Yes.

> do we need another Proc-wide fence between SuspendEE and starting GC? 

I do not think so. The coop threads are blocked during SuspendEE. The OS wait primitives act as fences.

Reply:
It feels like if we care about speculative reads happening before switching to coop, we should certainly worry about buffered writes happening after going to preemptive.

Reply:
I mean - if a thread does writes while in coop, switches to preemptive (unordered write on ARM) , GC starts and only then the writes are arriving...

Reply:
Hmm, you may be right. @Maoni0 ?

Reply:
the short answer is yes. but I wanted to give a long answer just so that we are all on the same page since suspension is such a tricky topic.

when Jan and I were first talking about this, I mentioned that I wasn't sure if a thread being switched out means a full barrier - I used the BGC thread case as an example and if it gets switched out and `g_TrapReturningThreads` is already false, we wanted to make sure there's a full barrier before it starts reading managed objects. now I believe when there's a context switch happening on the thread, it acts as a full barrier. so I don't think it's actually a problem even without the fix for the BGC thread case (for this purpose BGC threads are like normal managed threads). however for restart, there's a case that involves pinvoke and doesn't involve a context switch - a thread could happily be doing native stuff and running, and when it comes back to managed and `g_TrapReturningThreads` is already false, it can start working with managed objects right away, without any context switch at all. so I think my fix was needed for that case.

I believe the same thing applies to suspend - a thread was in coop mode, and now it's executing some native code for a while so it switched itself to preempt before starting the native work, and now a suspension happens and this thread is left alone but its writes may still be buffered since no cswitch needed to occur.

there's already a `FlushProcessWriteBuffers` in the suspension code path but it doesn't help with what we are talking about here at all. it seems like I should actually be adding a pair of `FlushProcessWriteBuffers` in GC before and after the GC work, ie, I move the `FlushProcessWriteBuffers` I added in Restart into the GC code and add one before GC starts. that way the intention is very clear.

Reply:
Suspension code seems to be the right place to have these barriers. Anything that calls suspension to access other thread's state will have these issues. 
That is in theory. In practice of all suspension users only GC/BGC might have tight enough timings to be affected and still very rarely.

Reply:
One way to look at this is that suspend/resume acts as mutual exclusion (lock) to accessing managed objects. On architectures with weak memory acquiring a lock does a read fence and releasing a lock does a write fence so that reads/writes do not "leak" outside of locked region.

In our suspension implementation there are two ways for a managed thread to take/release the "lock". - 1) By actually suspending/resuming its execution, no concerns here, that is a full fence. 2) By switching between coop/preemptive modes, which is basically a write into a field like [here](https://github.com/dotnet/runtime/blob/395fe0c5c20318eb788a030875fda515810f7b47/src/coreclr/src/vm/threads.h#L2005). Not having fences there makes sense since coop/preemptive switches are relatively frequent. However it means that EE suspend/release needs to do process-wide barrier. Otherwise, indeed, anyone who suspends EE and proceeds to read the shared state may find that some threads have "suspended" by simply leaving coop mode and their buffered writes from before that are still arriving. Symmetrical issue could exist on the resuming end too.

This is why I think it should not be a responsibility of GC to do these barriers. It should be a part of suspension/release, which implements the "lock".

Reply:
I'm not sure that you read what I wrote.

I was saying that I didn't actually think BGC threads had a problem, even without the fix I added. it's the interop that had a problem. 

and the reason why I proposed to have these in GC is because it's purely for reading/writing to objects which GC cares about.

Reply:
BGC starts with calling SuspendEE before it scans roots. Not sure if that is considered a part of BGC. Perhaps BGC prep? 

Without the fix SuspendEE does not guarantee that other cores have committed all their writes by the time it returns. I.E. some new roots may still be in write buffers. Adding FlushProcessWriteBuffers can ensure that GC sees them. But it would be simpler to add the call inside SuspendEE itself, rather where it is called. Nearly everything that calls SuspendEE will expect this guarantee.

Reply:
I mean - GC or BGC do not have a problem with this, not directly. The root cause is SuspendEE itself. Without flushing other cores' buffers we can't claim that EE is truly suspended.

Reply:
we were discussing the threads that were being suspended, not the one that calls SuspendEE. 

to sum up, I do believe we should flush the write buffers before GC starts as well. I should have thought about this when I was thinking about the change I made. so it's good you brought it up. thanks! I don't have a strong opinion on where to flushing other threads write buffers.